### ISSUE TITLE
Speech To Text: Response Always return Nothing-when 

### ISSUE THREAD

**Issue by RallyForecast** [googleapis/google-cloud-dotnet#6470]
The var Response always return NOTHING-when RecognitionAudio streams the byte array 
The pictures, explaining the errors - attached. The byte array with data from microphone has data(shown on the picture1) and return (var response) -> gets NOTHING. Though no errors happened.
Please help !
![05-17-21-1](https://user-images.githubusercontent.com/84359768/118575476-8fa4a500-b754-11eb-9d59-35884c2ed7d0.png)
![05-17-21-2](https://user-images.githubusercontent.com/84359768/118575478-8fa4a500-b754-11eb-857b-9bf83bbf517d.png)

var speech = SpeechClient.Create();
                var response = speech.Recognize(new RecognitionConfig()
                {
                    Encoding = RecognitionConfig.Types.AudioEncoding.Linear16,
                    SampleRateHertz = 16000,
                    LanguageCode = "en",
                }, RecognitionAudio.FromBytes(bytes.ToArray()));

                Console.WriteLine(response);

**Comment by jskeet** [comment]
Please attach a file containing the data you're sending - we can't tell what format the data is in, or whether it actually contains any speech.

**Comment by RallyForecast** [comment]
Here is the whole file. If you need - I can attach the whole project - pls let me know. Thank you

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using System.Diagnostics;
using Google.Cloud.Speech.V1;
using System.IO;
using System.Net;

namespace Tutorial
{
    public partial class Form1 : Form
    {
        public NAudio.Wave.BufferedWaveProvider waveBuffer = null; 
        public NAudio.Wave.BufferedWaveProvider waveProvider;

        string credential_path = @"C:\Tutorial\package.json";

        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            List<NAudio.Wave.WaveInCapabilities> sources = new List<NAudio.Wave.WaveInCapabilities>();

            for (int i = 0; i < NAudio.Wave.WaveIn.DeviceCount; i++)
            {
                sources.Add(NAudio.Wave.WaveIn.GetCapabilities(i));
            }

            sourceList.Items.Clear();

            foreach (var source in sources)
            {
                ListViewItem item = new ListViewItem(source.ProductName);
                item.SubItems.Add(new ListViewItem.ListViewSubItem(item, source.Channels.ToString()));
                sourceList.Items.Add(item);
            }
        }

        NAudio.Wave.WaveIn sourceStream = null;
        NAudio.Wave.DirectSoundOut waveOut = null;

        // Start Button
        private void button2_Click(object sender, EventArgs e)
        {
            if (sourceList.SelectedItems.Count == 0) return;

            int deviceNumber = sourceList.SelectedItems[0].Index;

            sourceStream = new NAudio.Wave.WaveIn();
            sourceStream.DeviceNumber = deviceNumber;/*44100*/
            sourceStream.WaveFormat = new NAudio.Wave.WaveFormat(16000,16, 1);

            NAudio.Wave.WaveInProvider waveIn = new NAudio.Wave.WaveInProvider(sourceStream);
            waveOut = new NAudio.Wave.DirectSoundOut();
            waveOut.Init(waveIn);
            sourceStream.StartRecording();
            // in order to handle the Stream :
            // https://stackoverflow.com/questions/34385261/from-naudio-wave-wavein-to-stream
            //waveOut.Play();

            sourceStream.DataAvailable += sourceStream_DataAvailable;
        }

        private void sourceStream_DataAvailable(object sender, NAudio.Wave.WaveInEventArgs e)
        {
            if (sourceStream == null)
                return;
            try
            {
                short[] audioData = new short[e.Buffer.Length / 2]; //this is your data! by default is in the short format
                Buffer.BlockCopy(e.Buffer, 0, audioData, 0, e.Buffer.Length);
                float[] audioFloat = Array.ConvertAll(audioData, x => (float)x); //I typically like to convert it to float for graphical purpose
               
                // OR :**********************=>

                byte[] bytes = new byte[e.Buffer.Length]; //recorded data is stored in this byte array
                Buffer.BlockCopy(e.Buffer, 0, bytes, 0, e.Buffer.Length);

                //start ************************************
                System.Environment.SetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS", credential_path);

                RecognitionAudio audio5 = null;

                using (Stream stream1 = new MemoryStream(bytes.ToArray())) // Any regular .NET stream
                {
                    audio5 = RecognitionAudio.FromStream(stream1);
                    SpeechClient client = SpeechClient.Create();
                    RecognitionConfig config = new RecognitionConfig
                    {
                        Encoding = RecognitionConfig.Types.AudioEncoding.Flac,
                        SampleRateHertz = 16000,
                        LanguageCode = LanguageCodes.Russian.Russia
                    };
                    RecognizeResponse response = client.Recognize(config, audio5);
                    Console.WriteLine(response);
                }
                
                //end   ************************************

            }
            catch (Exception exc)
            { //if some happens along the way...
            }
        }

        protected  string ConvertX(byte[] data)
        {
            char[] characters = data.Select(b => (char)b).ToArray();
            return new string(characters);
        }

       
        // Convert Array of Bytes to string
        // https://www.techiedelight.com/convert-byte-array-to-string-csharp/#:~:text=This%20post%20will%20discuss%20how%20to%20convert%20a,Encoding%20class%20%E2%80%93%20UTF8%2C%20Unicode%2C%20UTF32%2C%20ASCII%2C%20etc.

        protected string restoreArrayBytesToString(byte[] ar)
        {
            string outstr = "";
            byte[] bytes = Convert.FromBase64String("QUJDMTIz");
            Console.WriteLine("Byte Array is: " + String.Join(" ", bytes));

            string str = Convert.ToBase64String(bytes);
            Console.WriteLine("The String is: " + str);

            byte[] bytes2 = Encoding.ASCII.GetBytes("ABC123");
            Console.WriteLine("Byte Array is: " + String.Join(" ", bytes));

            string str2 = BytesToString(bytes);
            Console.WriteLine("The String is: " + str);

            return outstr;
        }

        static string BytesToString(byte[] bytes)
        {
            using (MemoryStream stream = new MemoryStream(bytes))
            {
                using (StreamReader streamReader = new StreamReader(stream))
                {
                    return streamReader.ReadToEnd();
                }
            }
        }
        private void button3_Click(object sender, EventArgs e)
        {
            if (waveOut != null)
            {
                waveOut.Stop();
                waveOut.Dispose();
                waveOut = null;
            }
            if (sourceStream != null)
            {
                sourceStream.StopRecording();
                sourceStream.Dispose();
                sourceStream = null;
            }
        }
        private void button4_Click(object sender, EventArgs e)
        {
            button3_Click(sender, e);
            this.Close();
        }
    }
}

**Comment by jskeet** [comment]
That's the *code* - but it's the *data* I was after. Before you send the data to the Speech service, please save it as a file, and attach that to the bug.

Note that it looks like you're currently sending it on the `DataAvailable` event - which I'd expect to happen very often, with a *tiny* piece of sound on each occurrence. That's not going to be enough to recognize speech in a standalone basis.
You can use streaming speech recognition to do that. There's some sample code for that [here](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/95b32e683ba534883b8a7f3c979deee101ba3678/speech/api/Recognize/InfiniteStreaming.cs)

**Comment by RallyForecast** [comment]
Thank you for sample code, but it is not complete  - could you please share also with the code for the Client - how to call it from the Console Application - the similar how it was done in Cloud Google for the Node.js and Java. Thank you again.

**Comment by jskeet** [comment]
If you clone the whole sample at [that commit](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/95b32e683ba534883b8a7f3c979deee101ba3678) you can run the [Recognize project](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/95b32e683ba534883b8a7f3c979deee101ba3678/speech/api/Recognize).

(The sample is no longer in the main branch, unfortunately, but if you clone the whole repo and then check out that commit, you should be fine.)

But the main point was to indicate that there's nothing wrong with the client library - it's just that you're using it incorrectly, attempting to recognize tiny snippets of sound as if they were complete files. I don't know that I'd recommend using the streaming API here unless you really need to - it's generally simpler if you can have a "start" button at which point you start recording and an "end" button at which point you stop recording, then send the whole recording at that point.

**Comment by RallyForecast** [comment]
Could I run the whole solution from regular VS2019 ? Or the examples need to be run only from Console Cloud Studio ? Could I run just the only Recognize solution from VS2019 ? I opened the Google Account, got .json file ... But the thing is that I want to use Stream, not Files(Start/Stop Buttons)... Usually there are no problems with running from github, but this time - I'm really in the stumb...Have No clew - how to run the class with Stream from Regular Console App...It should be some hints or examples how to do it, otherwise - it is useless samples...Also, not clear, why there are clear full examples with Python,Node,Java and nothing with C#...

**Comment by jskeet** [comment]
Yes, you can run it from VS2019. In the `Recognize` project debugging properties page (or in your Windows shell before starting Visual Studio), set the GOOGLE_APPLICATION_CREDENTIALS environment variable to refer to the service account JSON file you've downloaded. Specify whatever command line arguments you want (e.g. `listen` or `listen-infinite`), and make sure you've got the `Recognize` project as the startup project. (Or you can just run it from the command line, of course.)

The part of `Recognize.cs` that will be of most interest to you is the `StreamingRecognizeAsync` method. Within the `InfiniteStreaming.cs` file, the whole file is important - the point is that streaming recognition only works for a few minutes. If you only *need* a few minutes, that's fine, and the `StreamingRecognizeAsync` example is the one you should follow. If you need to be able to transcribe longer portions of text, that's where the complexities of `InfiniteStreaming.cs` are important.

I can't comment on why some languages have examples and others don't, other than to say that API teams tend to focus on the languages where they have more customers. There are relatively few of us writing C# libraries and samples. I will pass your feedback to the product team, that C# samples would be appreciated.

**Comment by RallyForecast** [comment]
Does it mean that stream will be breaking every time by 5 minutes pause and then restored automatically, getting new portion of data from some buffer ?

**Comment by RallyForecast** [comment]
Hi Jon,
I got it working. Thank you very much.
Next problem :) => it understands only English Speech.
When I set LanguageCode = "ru" - it appears "????" on the black screen .....
I tried "ru-RU", but still the same result.
Please let me know how to overcome this issue.
Thank you
Rally

**Comment by jskeet** [comment]
> When I set LanguageCode = "ru" - it appears "????" on the black screen .....

I very, very strongly suspect that that has nothing to do with the Speech API, and everything to do with how Cyrillic displays on your console. Two things to try:

- Create a console app which doesn't do anything with the Speech API, but just tries to print some Cyrillic text on the console.
- Run the Speech app in the debugger, and put a break point just before you write to the console. Look at what is about to be printed.

I expect that the first app will still show "???" in the console, whereas in the debugger when using the Speech app, you'll see Cyrillic in the debugger.

Assuming that's the case, please ask a question on Stack Overflow with all the details of your environment, as it's unrelated to any Google libraries. If that *isn't* the case, please create a new issue here (as it's no longer anything to do with your original problem).
